### PR TITLE
Config: Separate from constants

### PR DESCRIPTION
**Pull request by sentientwaffle**
## Background

Distinguish "config" from "constants":

- Configuration is anything that might be modified at build-time.
- Constants include:
  - Configuration (Constants are a superset of config).
  - Constants derived from configuration.
  - Universal constants (e.g. `sector_size`).

Distinguish between "process config" and "cluster config":
- _process config_ can be modified per-replica/client — different process do not need to agree, and a process can be safely restarted with a distinct process config.
- _cluster config_ must be identical on call replicas/clients — it factors in to the storage format and/or network protocol.

## Changes

- Rename `config.zig` to `constants.zig` (but we still use `const config = @import("constants.zig");` — I didn't want to make the diff even more unmanageable).
- A preset config can be chosen at build time. By default:
  - `main.zig` builds with the `development` config
  - Tests, fuzzers, and the VOPR build with the `test_min` config.
- Groove/Tree: Allow zero-sized value caches (https://github.com/tigerbeetledb/tigerbeetle/pull/301/commits/c53647432fe25fdbad7b440e71d54b705560d50b).
- Change the default value cache sizes: favor accounts, don't cache transfers.

There are currently 3 preset configs:

- `test_min` is a minimal configuration to allow deep testing of the LSM in-memory.
- `production`
- `development` is `production` with smaller caches (and a compatible storage format).

## Future work
(For a future PR since this one is functional but pretty giant already...)

- Rename `config` to `constants` everywhere.
- Support more granular config updates at build time (e.g. `-Dconfig-log-level=debug`).
- Update `tigerbeetle version --verbose` — it should probably only print config (not constants), and distinguish between "process" configs and "cluster" configs.
- https://github.com/tigerbeetledb/tigerbeetle/issues/295
- Fuzz/VOPR _multiple_ different configs instead of always using `test_min`.

### CODE REVIEW

File: build.zig [hunk 1]
```diff
@@ -24,12 +24,18 @@ pub fn build(b: *std.build.Builder) void {
...
 
+    options.addOption(
+        ConfigBase,
+        "config_base",
+        b.option(ConfigBase, "config", "Base configuration.") orelse .default,
+    );
+
     const tracer_backend = b.option(
-        TracerBackend,
+        config.TracerBackend,
```
Comment (jamii) [inline build.zig:34]: Now that TracerBackend is in config, we can use it in https://github.com/tigerbeetledb/tigerbeetle/blob/dec1160d09494d9853a166153b953b9f6d6cdc59//src/clients/node/src/node.zig#L32 too.

File: src/config.zig [hunk 2]
```diff
@@ -1,413 +1,198 @@
...
+    direct_io: bool,
+    direct_io_required: bool,
+    journal_iops_read_max: usize = 8,
+    journal_iops_write_max: usize = 8,
+    tick_ms: u63 = 10,
+    rtt_ms: u64 = 300,
+    rtt_multiple: u8 = 2,
+    backoff_min_ms: u64 = 100,
+    backoff_max_ms: u64 = 10000,
+    clock_offset_tolerance_max_ms: u64 = 10000,
```
Comment (jamii) [inline src/config.zig:55]: I haven't looked at how the clock sync works yet, but I'm suprised that different replicas can have different settings for it.

Comment (sentientwaffle) [reply src/config.zig:55]: You probably _shouldn't_ use different values, but we don't rely on clock sync for correctness, only availability. `ConfigProcess` is for things that are safe to change at "cluster runtime" — e.g. you could do a rolling replica restart to deploy a new value.

Comment (jamii) [reply src/config.zig:55]: :+1:

**Comment by jamii** [comment]
Looks good.

**Comment by sentientwaffle** [comment]
Oops, I need to fix the node client

**Comment by jamii** [comment]
Have you tried running tree_fuzz yet with the min config? I'd be curious to see how many lsm levels it can reach now.

**Comment by sentientwaffle** [comment]
I did! It found https://github.com/tigerbeetledb/tigerbeetle/pull/296 and https://github.com/tigerbeetledb/tigerbeetle/pull/300.

In the run I was debugging yesterday it reached level 4.

**Review by jamii — DISMISSED** [review]

**Comment by sentientwaffle** [comment]
I'm trying to figure out why the go client tests fail, then this will be good to go.

**Comment by jamii** [comment]
I need to learn to refresh the page to avoid stale CI results :S

**Comment by sentientwaffle** [comment]
Ignore this for a while; I botched a rebase...

File: .github/workflows/test_and_build.yml [hunk 3]
```diff
@@ -100,10 +100,10 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - run: ./scripts/install_zig.sh
-      - run: zig/zig build fuzz_vsr_superblock -- --seed 123
```
Comment (jamii) [inline .github/workflows/test_and_build.yml:103]: Should we put tree/forest fuzz in CI too?

Comment (sentientwaffle) [reply .github/workflows/test_and_build.yml:103]: I'll add the tree fuzzer in a follow-up PR. There is a bug in the forest fuzzer (revealed by the new minimal config) that makes it fail right now though.

**Review by jamii — APPROVED** [review]

**Pull request merged**